### ISSUE TITLE
Webpack can generate incorrect code if the original source defines async iterator in a class

### ISSUE THREAD

**Issue by YECHUNAN** [webpack/webpack#17565]
# Bug report

**What is the current behavior?**

Webpack 5.84.0 and newer can generate incorrect code when the source is an esm module and uses async iterator in some class definition.

**If the current behavior is a bug, please provide the steps to reproduce.**

Create two files, one called `my.esm.js` and the other one called `webpack-index.js`. Create a `webpack.config.js` as well.

Each file has the following code

my.esm.js
```javascript
var eP=[1,2],
Ky=class{
    constructor(){}
},
Io=class{
    constructor(t){if(!eP.includes(t.version))throw"invalid parquet version";}
    async*[Symbol.asyncIterator](){yield ""}
}
var SP=Io,w8={ParquetReader:SP};export{w8 as default};

```

webpack-index.js
```javascript
const parquetjs = require("./my.esm");

const reader = new parquetjs.default.ParquetReader(1);
```

webpack.config.js
```javascript
module.exports = (env, argv) => {
    const mode = argv.mode || "production";

    return {
        mode: mode,
        devtool: 'source-map',
        entry: './webpack-index.js',
        output: {
            path: __dirname + "/prebuilt-bundles",
            filename: (mode === 'production' ? 'webpack-bundle.js' : 'webpack-bundle.dev.js')
        },
        target: "web",
    }
}
```

Running webpack in production mode with the given source and config will reproduce the bug.

The generated code looks like this (only ~700 characters feel free to paste to a separate editor and toggle word wrap to read it). To quickly get to what's wrong, search for "null.includes".
```javascript
(()=>{var e={12:(e,r,t)=>{"use strict";t.r(r),t.d(r,{default:()=>o});var o={ParquetReader:class{constructor(e){if(!null.includes(e.version))throw"invalid parquet version"}async*[Symbol.asyncIterator](){yield""}}}}},r={};function t(o){var a=r[o];if(void 0!==a)return a.exports;var n=r[o]={exports:{}};return e[o](n,n.exports,t),n.exports}t.d=(e,r)=>{for(var o in r)t.o(r,o)&&!t.o(e,o)&&Object.defineProperty(e,o,{enumerable:!0,get:r[o]})},t.o=(e,r)=>Object.prototype.hasOwnProperty.call(e,r),t.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},new(t(12).ParquetReader)})();
//# sourceMappingURL=webpack-bundle.js.map
```

Class `Io` references a local variable `eP` that has array literal value [1,2]. In the generated code, the local variable `eP` is gone. The reference of `eP.includes` in class `Io` is translated to `null.includes` which will always throw when user runs the generated code.

**What is the expected behavior?**

The expected behavior is so that webpack keeps the local variable and use it at the right place.

**Other relevant information:**
webpack version:  5.84.0 and up to 5.88.2
Node.js version: 18.17.0
Operating System: Windows/Mac
Additional tools: N/A

**Additional context**

The code is from npm package @dsnp/parquetjs which we use in our product. The original file is too big so I trimmed it down to make it more feasible for troubleshooting. The problem seems to be very closely related to the async iterator defined in class `Io`. I tried tracing what webpack is doing. It turns out that when webpack walks through the definition of class `Io`, it attempts to evaluate the expression `Symbol.asyncIterator` and as a result flags the class as "not pure". Because of these changes added in 5.84.0 around this line in `InnerGraphPlugin.js`, since the class `Io` is flagged as "not pure", it will then walk through the body of its methods (constructor in this repro case) without calling `classWithTopLevelSymbol.set` to set the correct class scope. https://github.com/webpack/webpack/blob/main/lib/optimize/InnerGraphPlugin.js#L167
Basically, it incorrectly believes that the variable `eP` is referenced in class `Ky` which is defined above. Since class `Ky` is actually not used, both class `Ky` and the variable `eP` are wiped out during the optimization phase and results in the `null` value ended up where the variable should be referenced. If I delete the async iterator from class `Io` or simply move the definition of class `Ky` to be after the definition of `Io`, the generated code will become correct.

In general, I think the scenario is too subtle for developers to keep in mind while using webpack. I think the fix should be in webpack instead of in the client code. I made a PR proposing a fix for the aforementioned issue.

**Comment by YECHUNAN** [comment]
Ping.
